---
title: "Orthology Results: Assessing the initial orthogroups (homogroups) 155 metazoans"
author: "Dean Mckeown"
date: "`r Sys.Date()`"
output: 
  html_document:
    output_file: "results/homogroups.html"
    toc: true
    toc_float: TRUE
    toc_depth: 4
    number_sections: true
    theme: united
    highlight: tango
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(echo = TRUE, fig.path = "results/")

getwd()

```

```{css, echo=FALSE}
.scroll-200 {
  max-height: 200px;
  overflow-y: auto;
  background-color: inherit;
}
```

___



```{r Preparation, include = FALSE}
source('R/hg_data_prep.R')
suppressPackageStartupMessages(suppressWarnings(source('R/homogroup-functions.R')))


################################################################################
# INPUTS
################################################################################

################################################################################
# General inputs
################################################################################
# Colour palette (must match taxonomy in samplesheet1)

SUPERGROUP_COLS <- c(
  Choanoflagellata = "#A6CEE3",
  Cnidaria = "#1F78B4",
  Ctenophora = "#B2DF8A",
  Deuterostomia = "#33A02C",
  Filasterea = "#FB9A99",
  Placozoa = "#E31A1C",
  Porifera = "#FF7F00",
  Protostomia = "#CAB2D6",
  Teretosporea = "#6A3D9A"
)


OG_CALLER_COLS <- c(
  orthofinder = "#D95F02",
  broccoli = "#1B9E77"
)

sg_axis_lab <- main_seq_long %>%
  select(id, supergroup) %>%
  as.data.frame() %>%
  unique() %>%
  ggplot(aes(id, 1, fill = supergroup)) +
  geom_col(width = 1, position = "identity") +
  scale_fill_manual(values = SUPERGROUP_COLS) +
  coord_cartesian(expand = FALSE) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 16),
        axis.text.y = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_blank()) +
  theme(legend.position = "none")


################################################################################
# OUTPUTS
################################################################################
# And where summarized results should be saved.
OUTPUT_DIR <- './results/'
dir.create(OUTPUT_DIR, showWarnings = F)

```


```{r Species tree}
tree <- read.tree(SPECIES_TREE)

tree_meta <- list()

tree_meta$tips <- main_seq_long %>%
  select(
    id,
    supergroup
  ) %>%
  unique()

# prepare data for heatmap
# per geneclass
tree_meta$hm_total_genes_in_id_by_geneclass <- main_seq_long %>%
  select(id, most_common_pfam_hmm_name, clean_parent_seq) %>%
  unique() %>%
  group_by(id, most_common_pfam_hmm_name) %>%
  summarise(values = n_distinct(clean_parent_seq), .groups = 'drop') %>%
  pivot_wider(names_from = most_common_pfam_hmm_name, values_from = values) %>%
  column_to_rownames("id") %>%
  replace(is.na(.), 0)

# per OG
tree_meta$hm_total_genes_in_id_by_og <- main_seq_long %>%
  select(id, Orthogroup, clean_parent_seq) %>%
  unique() %>%
  group_by(id, Orthogroup) %>%
  summarise(values = n_distinct(clean_parent_seq), .groups = 'drop') %>%
  pivot_wider(names_from = Orthogroup, values_from = values) %>%
  column_to_rownames("id") %>%
  replace(is.na(.), 0)

sp_tree <- tree %>%
  ggtree(aes(color = supergroup),
    linewidth = 1.0,
    layout = "rectangular"
  ) %<+% tree_meta$tips +
  geom_tiplab() +
  scale_color_manual(values = SUPERGROUP_COLS)

sp_tree <- gheatmap(sp_tree, 
  tree_meta$hm_total_genes_in_id_by_geneclass,
  offset = 8,
  width = 2,
  colnames_angle = 90,
  colnames_offset_y = -1,
  hjust = 1,
  low = "white",
  high = "blue",
  color = "white"
  )

sp_tree +
  ggtree::vexpand(.1, -1) +
  coord_cartesian(clip = "off")

# Save the tree and metadata for shiny
save(tree, tree_meta, file = "data/tree_and_metadata.RData")


```
## Orthogroup Homogeneity {.tabset}
```{r Orthogroup homogeneity}
# data reshape
og_meta <- main_seq_long %>%
  select(
    id,
    supergroup,
    Orthogroup,
    OG_source,
    cogeqc_hscore,
    cogeqc_hscore_scaled_all,
    cogeqc_hscore_scaled_self,
    total_seqs,
    total_parent,
    total_ids,
    total_supergroups,
    most_common_pfam_hmm_name,
    total_most_common_pfam_hmm_name,
    percent_most_common_pfam_hmm_name,
    vs_OG_source,
    Best_Jaccard,
    Mean_Jaccard
  ) %>%
  rename(og_size = total_seqs) %>%
  unique() %>%
  collapse_col(id, c("OG_source", "Orthogroup"), ",", unique_values = TRUE) %>%
  unique() %>%
  collapse_col(supergroup, c("OG_source", "Orthogroup"), ",", unique_values = TRUE) %>%
  unique()

# plot by tools
vn_cogeqc_hssc_tools <- og_meta %>% ggviolin(
    y = "cogeqc_hscore_scaled_all", x = "OG_source", 
    trim = TRUE, 
    add = "jitter",
    fill = "OG_source"
  ) +
    scale_fill_manual(values = OG_CALLER_COLS) +
    labs(y = "Scaled homogeneity scores", x = "Source of orthogroups",
         title = "Distribution of mean homogeneity scores for orthogroups") +
    theme(plot.subtitle = ggtext::element_markdown())

# plot by species, color tools
vn_cogeqc_hssc_species <- og_meta %>%
separate_longer_delim(id, delim = ",") %>%
separate_longer_delim(supergroup, delim = ",") %>%
unique() %>%
ggplot(aes(x = id, y = cogeqc_hscore_scaled_all, fill = OG_source)) +
  geom_violin(trim = TRUE, position = position_dodge(1)) +
  scale_fill_manual(values = OG_CALLER_COLS) +
  labs(y = "Scaled homogeneity scores", x = "Species",
       title = "Distribution of mean homogeneity scores for orthogroups") +
  theme(plot.subtitle = ggtext::element_markdown())

# hack to label axis with supergroup
#vn_cogeqc_hssc_species + sg_axis_lab + plot_layout(ncol = 1, nrow = 2, heights = c(20, 1))

```

### OG Homogeneity by Tools
```{r}
vn_cogeqc_hssc_tools
```

### OG Homogeneity by Species
```{r}
vn_cogeqc_hssc_species
```
---
## Orthogroup Stats {.tabset}
```{r Basic OG stats}
# reshape data to be OG centric

sp_ogs_coghs_vs_size_vs_sgs <- og_meta %>%
  ggplot( aes(x = og_size,
  y = cogeqc_hscore_scaled_all,
  size = total_supergroups,
  color = OG_source)) +
  geom_point(alpha=0.7) +
  scale_size(range = c(0, 10)) +
  scale_color_manual(values = OG_CALLER_COLS) +
  theme(legend.position = "none") +
  theme_bw()

vn_size_tools <- og_meta %>%
    ggviolin(y = "og_size", x = "OG_source", 
    trim = TRUE, 
    add = "jitter",
    fill = "OG_source"
  ) +
    scale_fill_manual(values = OG_CALLER_COLS) +
    labs(y = "Num genes in OG", x = "Source of orthogroups") +
    theme(plot.subtitle = ggtext::element_markdown()) +
    scale_y_continuous(trans = 'log10', labels = scales::comma)


```
### Orthogroup homogeneity score vs OG size
```{r}
sp_ogs_coghs_vs_size_vs_sgs
```
### Orthogroup size by tool
```{r}
vn_size_tools
```
---

```{r Algorithm Comparison}
# Create lists of sequences associated with each algorithm and OG
compare_callers <- main_ogs %>%
  split(.$OG_source) %>%
  lapply(function(x) split(x, x$Orthogroup)) %>%
  lapply(function(group) {
    lapply(group, function(x) unique(strsplit(x$clean_parent_seqs, ",")))
  })

# Initialize an empty list to store the Jaccard similarity matrices
jaccard_results <- list()

# Perform a comparison and store the result in the list with a meaningful name
jaccard_results$of_vs_br <- calculate_jaccard_similarity(compare_callers[["orthofinder"]], compare_callers[["broccoli"]])
jaccard_results$br_vs_of <- calculate_jaccard_similarity(compare_callers[["broccoli"]], compare_callers[["orthofinder"]])
jaccard_results$of_vs_of <- calculate_jaccard_similarity(compare_callers[["orthofinder"]], compare_callers[["orthofinder"]])
jaccard_results$br_vs_br <- calculate_jaccard_similarity(compare_callers[["broccoli"]], compare_callers[["broccoli"]])

# Assuming jaccard_results is a list with matrices for each pair
jaccard_metrics <- list()
jaccard_metrics$of_vs_br <- calculate_jaccard_metrics(jaccard_results$of_vs_br, "orthofinder", "broccoli")
jaccard_metrics$br_vs_of <- calculate_jaccard_metrics(jaccard_results$br_vs_of, "broccoli", "orthofinder")
jaccard_metrics$br_vs_br <- calculate_jaccard_metrics(jaccard_results$br_vs_br, "broccoli", "broccoli")
jaccard_metrics$of_vs_of <- calculate_jaccard_metrics(jaccard_results$of_vs_of, "orthofinder", "orthofinder")

jaccard_metrics_all <- bind_rows(jaccard_metrics)

jaccard_metrics_all_long <- jaccard_metrics_all %>%
  gather(key = "Metric", value = "Value", -tool1, -tool2)

# Create the heatmap using ggplot2
jaccard_metrics_heat_callers <- ggplot(jaccard_metrics_all_long, aes(x = tool1, y = tool2, fill = Value)) + 
  geom_tile() +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", midpoint = 0.5) +
  facet_wrap(~Metric, scales = "free") +
  theme_minimal() +
  labs(
    title = "Heatmaps of Metrics by Tool Pair",
    x = "Tool 1", y = "Tool 2", fill = "Value"
  ) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.text = element_text(size = 12))

jaccard_metrics_heat_callers

```

```{r Taxonomic Distribution}
upset_input <- list()
# make matrix for supergroups
upset_input$supergroup$orthofinder <- create_presence_matrix(filter(main_deflines, OG_source == "orthofinder"), supergroup, Orthogroup)

upset_input$supergroup$broccoli <- create_presence_matrix(filter(main_deflines, OG_source == "broccoli"), supergroup, Orthogroup)

upset_input$supergroups <- upset_input$supergroup %>% bind_rows()

supergroups <- colnames(upset_input$supergroups)[-1]


# make matrix for pfam_hmm_names gene classes
main_deflines_geneclasses <- main_deflines %>% separate_longer_delim(pfam_hmm_names, delim = ",") %>%
  select(Orthogroup, pfam_hmm_names, parent_seq, OG_source) %>%
  unique() %>%
  select(Orthogroup, pfam_hmm_names, OG_source) %>%
  mutate(pfam_hmm_names = coalesce(pfam_hmm_names, "None"))


upset_input$geneclass$orthofinder <- create_presence_matrix(filter(main_deflines_geneclasses, OG_source == "orthofinder"), pfam_hmm_names, Orthogroup)

upset_input$geneclass$broccoli <- create_presence_matrix(filter(main_deflines_geneclasses, OG_source == "broccoli"), pfam_hmm_names, Orthogroup)

upset_input$geneclasses <- upset_input$geneclass %>% bind_rows()

geneclasses <- colnames(upset_input$geneclasses)[-1]

# bring the mutiple sets together
upset_input_all <- left_join(upset_input$supergroups, upset_input$geneclasses, by = "Name")

# join in metadata
og_metadata_for_upset <- main_ogs %>%
  select(
    Orthogroup,
    OG_source
  ) %>%
  unique()

upset_input_all <- upset_input_all %>% 
  left_join(og_metadata_for_upset, by = c("Name" = "Orthogroup"))

upset_supergroups <- upset(upset_input_all,
  supergroups,
  annotations = list(
      'Orthogroup caller'=(
          ggplot(mapping=aes(fill=OG_source))
          + geom_bar(stat='count', position='fill')
          + scale_y_continuous(labels=scales::percent_format())
          + scale_fill_manual(values = OG_CALLER_COLS)
          + ylab('Orthogroup caller')
      )
  ),
)

upset_geneclasses <- upset(upset_input_all,
  geneclasses,
  annotations = list(
      'Orthogroup caller'=(
          ggplot(mapping=aes(fill=OG_source))
          + geom_bar(stat='count', position='fill')
          + scale_y_continuous(labels=scales::percent_format())
          + scale_fill_manual(values = OG_CALLER_COLS)
          + ylab('Orthogroup caller')
      )
  ),
)


```