---
title: "Orthology Results: Assessing the initial orthogroups (homogroups) 155 metazoans"
author: "Dean Mckeown"
date: "`r Sys.Date()`"
output: 
  html_document:
    output_file: "../results/homogroups.html"
    toc: true
    toc_float: TRUE
    toc_depth: 4
    number_sections: true
    theme: united
    highlight: tango
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.path = "../results/")
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

```{css, echo=FALSE}
.scroll-200 {
  max-height: 200px;
  overflow-y: auto;
  background-color: inherit;
}
```

___

## Preparation
We will begin by creating a variable that specifies where all workflow outputs are stored. Additionally we will source the script containing helper functions, and setting to variable some colors that we'll use throughout. 
### Inputs needed

/users/asebe/dmckeown/projects/crg-bcaortho/project-ioo_mz155/results/orthofinder_results/orthofinder_mcl/Orthogroups.tsv
  * **Pipeline generated, for any orthocaller** - which homogroups contain what sequences

/users/asebe/dmckeown/projects/crg-bcaortho/project-ioo_mz155/results/prefilter/initial/defline_info/*.csv
  * **Pipeline generated, for any orthocaller** - hmmsearch assignment for each sequence

/users/asebe/dmckeown/projects/crg-bcaortho/project-ioo_mz155/results/orthofinder_results/orthofinder_mcl/Orthogroups.GeneCount.tsv
  * **Pipeline generated, for any orthocaller** - just Orthogroups.tsv just totals instead of deflines

/users/asebe/xgraubove/genomes/annotation_functional/*_long.pep.pfamscan.csv
  * **Precalculated functional annotations**, needs gene id, PFAM accession for OG homogeneity assessment by cogeqc

### Inputs that MIGHT be needed, but currently only available for orthofinder
/users/asebe/dmckeown/projects/crg-bcaortho/project-ioo_mz155/results/orthofinder_mcl/complete_dataset/Results_Inflation_2.5/Comparative_Genomics_Statistics/Statistics_PerSpecies.tsv
  * Basic stats of ortholog size etc by species

/users/asebe/dmckeown/projects/crg-bcaortho/project-ioo_mz155/results/orthofinder_mcl/complete_dataset/Results_Inflation_2.5/Comparative_Genomics_Statistics/Orthogroups_SpeciesOverlaps.tsv
  * matrix of all samples vs all - might need to hunt down OF code
  * the number of orthogroups shared between each species-pair as a square matrix.

```{r Preparation, results = FALSE}
suppressPackageStartupMessages(suppressWarnings(source('R/packages.R')))
suppressPackageStartupMessages(suppressWarnings(source('R/homogroup-functions.R')))
source('R/arcadia-color-gradients.R')

# INPUTS
# Specify where the workflow outputs are stored (relative paths unless indicated)

RESULT_DIR  <-  "../project-ioo_mz155/results/orthofinder_results/"
#SAMPLESHEET <- "/users/asebe/dmckeown/projects/crg-bcaortho/project-ioo_mz155/samplesheet1.csv"
SAMPLESHEET <- "../project-ioo_mz155/samplesheet1.csv"

STATS_PER_SPP_TSV <- "../project-ioo_mz155/results/orthofinder_results/orthofinder_mcl/OrthoFinder/Results_Inflation_2.5/Comparative_Genomics_Statistics/Statistics_PerSpecies.tsv"
GENE_COUNTS_TSV <- "../project-ioo_mz155/results/orthofinder_results/orthofinder_mcl/Orthogroups.GeneCount.tsv"

# If you did not provide an external tree then it will be this:
#SPECIES_TREE <- paste0(RESULT_DIR, 'speciesrax/species_trees/inferred_species_tree.newick')
# Otherwise, give the path to the tree you provided
#SPECIES_TREE <- '/users/asebe/gzolotarov/projects/2021_TFevol/metazoan_tf_evol_2022/030523_phylogenies/data_annotation/species_tree.newick'
SPECIES_TREE <- '../../../../gzolotarov/projects/2021_TFevol/metazoan_tf_evol_2022/030523_phylogenies/data_annotation/species_tree.newick'

# OUTPUTS
# And where summarized results should be saved.
OUTPUT_DIR <- './results/'
dir.create(OUTPUT_DIR, showWarnings = F)
dir.create(paste0(OUTPUT_DIR, 'orthofinder'), showWarnings = F)

# OTHER PARAMETERS
# Set colours to plot supergroups (taxonomy field in samplesheet1)

SUPERGROUP_COLS <- c(
  Choanoflagellata = "#BEBADA",
  Cnidaria = "#FB8072",
  Ctenophora = "#80B1D3",
  Deuterostomia = "#8DD3C7",
  Filasterea = "#FDB462",
  Placozoa = "#B3DE69",
  Porifera = "#FCCDE5",
  Protostomia = "#FFFFB3",
  Teretosporea = "#D9D9D9"
)

```
___

## Gene Family Summaries {.tabset}
Now, let's go ahead and summarize the results from OrthoFinder. To start, this will just include some high-level summaries - number of species per gene family, gene copy number per gene family, etc.

```{r Gene Family Summary Stats, fig.asp = .95}
# Summarize gene family content, both with respect to number of species 
# included in each gene family, and the per-species mean copy number
summ_genefam_composition(results_dir = RESULT_DIR, show_plots = T, out_dir = paste0(OUTPUT_DIR, 'orthofinder/'))

# And get orthogroup summary statistics per species
per_spp_of_stats <-
  get_per_spp_ofinder_stats(tree_fpath = SPECIES_TREE,
                        samplesheet_fpath = SAMPLESHEET,
                        results_dir = RESULT_DIR, grp_name = 'Supergroup',
                        out_dir = paste0(OUTPUT_DIR, 'orthofinder/'),
                        tip_grp_cols = SUPERGROUP_COLS, 
                        count_cols = arcadia_cividis,
                        prop_cols = arcadia_viridis,
                        og_stats_perspp_fpath = STATS_PER_SPP_TSV)

# Lastly, get the per-species orthogroup counts/copy numbers
og_counts <- get_per_spp_og_counts(results_dir = RESULT_DIR, gene_count_tsv = GENE_COUNTS_TSV)
```

___

## Homogroup Homogeneity
```{r Homogroup homogeneity}
df_main <- prepare_all_data(samplesheet_fpath = SAMPLESHEET)

cogeqc_input_of <- prepare_cogeqc_inputs(
  input_df = df_main,
  orthogroup_col = Orthogroup_Orthofinder,
  species_col = id,
  gene_col = clean_seq,
  annot_col = hmm_acc,
  annot_filter_col = type,
  annot_filter_value = "Domain"
)

cogeqc_input_br <- prepare_cogeqc_inputs(
  input_df = df_main,
  orthogroup_col = Orthogroup_Broccoli,
  species_col = id,
  gene_col = clean_seq,
  annot_col = hmm_acc,
  annot_filter_col = type,
  annot_filter_value = "Domain"
)
# Calculate mean orthogroup homogeneity scores across species
og_assessment_of <- assess_orthogroups(cogeqc_input_of[["orthogroups"]], cogeqc_input_of[["annotations"]])
og_assessment_br <- assess_orthogroups(cogeqc_input_br[["orthogroups"]], cogeqc_input_br[["annotations"]])


og_hscores_of <- get_cogeqc_hscores(cogeqc_input_of[["orthogroups"]], cogeqc_input_of[["annotations"]])
og_hscores_br <- get_cogeqc_hscores(cogeqc_input_br[["orthogroups"]], cogeqc_input_br[["annotations"]])


# Example call
cogeqc_outputs <- compare_homogeneity_scores(
  og_hscores_of,
  og_hscores_br,
  comps = list(c("orthofinder", "broccoli"))
)

ggsave(cogeqc_outputs[["og_hscores_scaled_violin"]], filename = paste0(OUTPUT_DIR, "og_homogeneity_by_source.png"), 
           height = 6, width = 12, dpi = 300)



```

---

## Algorithm Comparison
```{r Algorithm Comparison}
# Create lists of sequences associated with each algorithm
comp_ogs_of <- split(df_main$clean_seq, df_main$Orthogroup_Orthofinder)
comp_ogs_of <- lapply(comp_ogs_of, unique)

comp_ogs_br <- split(df_main$clean_seq, df_main$Orthogroup_Broccoli)
comp_ogs_br <- lapply(comp_ogs_br, unique)

# Initialize an empty list to store the Jaccard similarity matrices
jaccard_results <- list()

# Perform a comparison and store the result in the list with a meaningful name
jaccard_results$of_vs_br <- calculate_jaccard_similarity(comp_ogs_of, comp_ogs_br)
jaccard_results$of_vs_of <- calculate_jaccard_similarity(comp_ogs_of, comp_ogs_of) # self of
jaccard_results$br_vs_br <- calculate_jaccard_similarity(comp_ogs_br, comp_ogs_br) # self br



# Define the function to calculate the metrics for a matrix
calculate_jaccard_metrics <- function(data, tool1_name, tool2_name) {
  data_matrix <- as.matrix(data)
  all_values <- as.vector(data_matrix)
  total_ogs <- nrow(data_matrix) + ncol(data_matrix) - 1

  rows_with_one <- sum(apply(data_matrix, 1, function(row) any(row == 1, na.rm = TRUE)))
  cols_with_one <- sum(apply(data_matrix, 1, function(col) any(col == 1, na.rm = TRUE)))
  total_ogs_with_one = rows_with_one + cols_with_one - 1
  portion_ones <- (total_ogs_with_one / total_ogs)

  max_jaccard_per_column <- apply(data_matrix, 2, max, na.rm = TRUE)
  mean_best_jaccard_index <- mean(max_jaccard_per_column, na.rm = TRUE)

  # Create a data frame with the results
  result <- data.frame(
    tool1 = tool1_name,
    tool2 = tool2_name,
    Portion_of_OGs_identical_pw = portion_ones,
    Mean_Best_Jaccard = mean_best_jaccard_index
  )
  
  return(result)
}


# Assuming jaccard_results is a list with matrices for each pair
metrics_of_vs_br <- calculate_jaccard_metrics(jaccard_results$of_vs_br, "OrthoFinder", "Broccoli")
metrics_br_vs_br <- calculate_jaccard_metrics(jaccard_results$br_vs_br, "Broccoli", "Broccoli")
metrics_of_vs_of <- calculate_jaccard_metrics(jaccard_results$of_vs_of, "OrthoFinder", "OrthoFinder")

metrics_all <- rbind(metrics_of_vs_br, metrics_br_vs_br, metrics_of_vs_of)

metrics_all_long <- metrics_all %>%
  gather(key = "Metric", value = "Value", -tool1, -tool2)

# Create the heatmap using ggplot2
ggplot(metrics_all_long, aes(x = tool1, y = tool2, fill = Value)) +
  geom_tile() +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", midpoint = 0.5) + 
  facet_wrap(~Metric, scales = "free") +
  theme_minimal() +
  labs(
    title = "Heatmaps of Metrics by Tool Pair",
    x = "Tool 1", y = "Tool 2", fill = "Value"
  ) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.text = element_text(size = 12))


```

```{r Taxonomic Distribution}
presence_absence_matrix <- df_main %>%
  select(Orthogroup_Orthofinder, supergroup) %>%
  distinct() %>%
  mutate(Presence = 1) %>%
  spread(key = supergroup, value = Presence, fill = 0) %>%
  rename(
    Name = Orthogroup_Orthofinder
  )


create_presence_matrix <- function(df, group_col, value_col) {
  df %>%
    select({{ value_col }}, {{ group_col }}) %>%  # Select the specified columns
    distinct() %>%  # Remove duplicate combinations
    mutate(Presence = 1) %>%  # Mark presence with 1
    spread(key = {{ group_col }}, value = Presence, fill = 0) %>%  # Convert to wide format
    rename(Name = {{ value_col }})  # Rename the value column to 'Name' for clarity
}

# Example usage with your data frame (df_main)
presence_absence_matrix_of <- create_presence_matrix(df_main, supergroup, Orthogroup_Orthofinder)
presence_absence_matrix_br <- create_presence_matrix(df_main, supergroup, Orthogroup_Broccoli)


```