---
title: "Orthology Results: Assessing the initial orthogroups (homogroups) 155 metazoans"
author: "Dean Mckeown"
date: "`r Sys.Date()`"
output: 
  html_document:
    output_file: "../results/homogroups.html"
    toc: true
    toc_float: TRUE
    toc_depth: 4
    number_sections: true
    theme: united
    highlight: tango
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.path = "../results/")
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

```{css, echo=FALSE}
.scroll-200 {
  max-height: 200px;
  overflow-y: auto;
  background-color: inherit;
}
```

___

## Preparation
We will begin by creating a variable that specifies where all workflow outputs are stored. Additionally we will source the script containing helper functions, and setting to variable some colors that we'll use throughout. 
### Inputs needed

/users/asebe/dmckeown/projects/crg-bcaortho/project-ioo_mz155/results/orthofinder_results/orthofinder_mcl/Orthogroups.tsv
  * **Pipeline generated, for any orthocaller** - which homogroups contain what sequences

/users/asebe/dmckeown/projects/crg-bcaortho/project-ioo_mz155/results/prefilter/initial/defline_info/*.csv
  * **Pipeline generated, for any orthocaller** - hmmsearch assignment for each sequence

/users/asebe/dmckeown/projects/crg-bcaortho/project-ioo_mz155/results/orthofinder_results/orthofinder_mcl/Orthogroups.GeneCount.tsv
  * **Pipeline generated, for any orthocaller** - just Orthogroups.tsv just totals instead of deflines

/users/asebe/xgraubove/genomes/annotation_functional/*_long.pep.pfamscan.csv
  * **Precalculated functional annotations**, needs gene id, PFAM accession for OG homogeneity assessment by cogeqc

### Inputs that MIGHT be needed, but currently only available for orthofinder
/users/asebe/dmckeown/projects/crg-bcaortho/project-ioo_mz155/results/orthofinder_mcl/complete_dataset/Results_Inflation_2.5/Comparative_Genomics_Statistics/Statistics_PerSpecies.tsv
  * Basic stats of ortholog size etc by species

/users/asebe/dmckeown/projects/crg-bcaortho/project-ioo_mz155/results/orthofinder_mcl/complete_dataset/Results_Inflation_2.5/Comparative_Genomics_Statistics/Orthogroups_SpeciesOverlaps.tsv
  * matrix of all samples vs all - might need to hunt down OF code
  * the number of orthogroups shared between each species-pair as a square matrix.

```{r Preparation, results = FALSE}
suppressPackageStartupMessages(suppressWarnings(source('R/packages.R')))
suppressPackageStartupMessages(suppressWarnings(source('R/homogroup-functions.R')))
source('R/arcadia-color-gradients.R')

################################################################################
# INPUTS
################################################################################

################################################################################
# Pipeline run inputs
################################################################################
RESULT_DIR  <-  "/no_backup/asebe/dmckeown/test_br_array_vs_br/results/"
SAMPLESHEET <- "../inputs/samplesheet1-10.csv"
PIPELINE_RUN_NAME <- "test_br_array_vs_br" # we will use this later to identify runs if comparing them this way

#SPECIES_TREE <- paste0(RESULT_DIR, 'speciesrax/species_trees/inferred_species_tree.newick') # internal speciesrax tree will be here
SPECIES_TREE <- '../../../../gzolotarov/projects/2021_TFevol/metazoan_tf_evol_2022/030523_phylogenies/data_annotation/species_tree.newick'

################################################################################
# Orthogroup caller specific stuff (subworkflow i.e. orthofinder, broccoli)
################################################################################

OGS_TSV_PATH_OF <- paste0(RESULT_DIR, "orthofinder_results/orthofinder_mcl/Orthogroups.tsv")
OGS_TSV_PATH_BR <- paste0(RESULT_DIR, "broccoli_array_results/broccoli/Orthogroups.tsv")

################################################################################
# General inputs
################################################################################
# Colour palette (must match taxonomy in samplesheet1)

SUPERGROUP_COLS <- c(
  Choanoflagellata = "#BEBADA",
  Cnidaria = "#FB8072",
  Ctenophora = "#80B1D3",
  Deuterostomia = "#8DD3C7",
  Filasterea = "#FDB462",
  #Placozoa = "#B3DE69",
  Porifera = "#FCCDE5",
  Protostomia = "#FFFFB3"
  #Teretosporea = "#D9D9D9"
)

LONG_PEP_PFAMSCAN_CSVS <- list.files("../../../../xgraubove/genomes/annotation_functional/", pattern = "*_long.pep.pfamscan.csv", full.names = TRUE)

GENE_FAMILIES_SEARCH_INFO <- "../../../../climate-adaptation-corals-ops/results_annotation/data/gene_families_searchinfo.csv"

################################################################################
# OUTPUTS
################################################################################
# And where summarized results should be saved.
OUTPUT_DIR <- './results/'
dir.create(OUTPUT_DIR, showWarnings = F)
dir.create(paste0(OUTPUT_DIR, 'orthofinder'), showWarnings = F)

```
___

## Homogroup Homogeneity
```{r Homogroup homogeneity}

# Get functional annotations
samplesheet <- read.delim(SAMPLESHEET, sep = ",")
species_list <- unique(samplesheet$id)

# Filter the file paths based on species in the orthogroups
LONG_PEP_PFAMSCAN_CSVS_FILT <- LONG_PEP_PFAMSCAN_CSVS[sapply(LONG_PEP_PFAMSCAN_CSVS, function(x) {
  any(sapply(species_list, function(species) grepl(species, x)))
})]

pfam_header <- c("seq_id", "alignment_start", "alignment_end", "envelope_start", "envelope_end", 
                  "hmm_acc", "hmm_name", "type", "hmm_start", "hmm_end", "hmm_length", 
                  "bit_score", "E_value", "significance", "clan")

# Read the filtered CSV files
functional_annotations <- lapply(LONG_PEP_PFAMSCAN_CSVS_FILT, clean_pfam_scan_file, header = pfam_header)
functional_annotations <- do.call(rbind, functional_annotations)

# Path to save the functional_annotations RDS file
all_data_path <- "data/all_data.rds"

  # Check if all three RDS files exist
  if (file.exists(all_data_path)) {
    cat(paste("File", all_data_path, "exists - loading data from it...\n"))
    # If all RDS files exist, load them
    all_data <- readRDS(all_data_path)
    return(all_data)

  } else {
  cat(paste("File", all_data_path, "does not exist - generating it...\n"))
  # repeat function call for each orthogroup caller to include
  all_data <- list()

    all_data[["orthofinder"]] <- prepare_all_data(
      results_dir = RESULT_DIR,
      samplesheet_fpath = SAMPLESHEET,
      orthogroups_path = OGS_TSV_PATH_OF,
      functional_annotations = functional_annotations,
      OG_source = "orthofinder" 
    )

    all_data[["broccoli"]] <- prepare_all_data(
      results_dir = RESULT_DIR,
      samplesheet_fpath = SAMPLESHEET,
      orthogroups_path = OGS_TSV_PATH_BR,
      functional_annotations = functional_annotations,
      OG_source = "broccoli" 
    )

  }

cogeqc_input <- list()

cogeqc_input[["orthofinder"]] <- prepare_cogeqc_inputs(
  input_df = all_data[["orthofinder"]],
  species_col = id,
  gene_col = clean_seq,
  annot_col = hmm_acc,
  annot_filter_col = type,
  annot_filter_value = "Domain"
)

cogeqc_input[["broccoli"]] <- prepare_cogeqc_inputs(
  input_df = all_data[["broccoli"]],
  species_col = id,
  gene_col = clean_seq,
  annot_col = hmm_acc,
  annot_filter_col = type,
  annot_filter_value = "Domain"
)
# Calculate mean orthogroup homogeneity scores across species
og_assessment <- list()
og_assessment[["orthofinder"]] <- assess_orthogroups(cogeqc_input[["orthofinder"]][["orthogroups"]], cogeqc_input[["orthofinder"]][["annotations"]])
og_assessment[["broccoli"]] <- assess_orthogroups(cogeqc_input[["broccoli"]][["orthogroups"]], cogeqc_input[["broccoli"]][["annotations"]])

og_hscores <- list()
og_hscores[["orthofinder"]] <- get_cogeqc_hscores(cogeqc_input[["orthofinder"]][["orthogroups"]], cogeqc_input[["orthofinder"]][["annotations"]])
og_hscores[["broccoli"]] <- get_cogeqc_hscores(cogeqc_input[["broccoli"]][["orthogroups"]], cogeqc_input[["broccoli"]][["annotations"]])

# Plot COGEQC
cogeqc_outputs <- compare_homogeneity_scores(
  og_hscores[["orthofinder"]],
  og_hscores[["broccoli"]],
  comps = list(c("orthofinder", "broccoli"))
)

ggsave(cogeqc_outputs[["og_hscores_scaled_violin"]], filename = paste0(OUTPUT_DIR, "og_homogeneity_by_source.png"), 
           height = 6, width = 12, dpi = 300)

# Add cogeqc scores to all_data
if (!file.exists(all_data_path)) {
  all_data[["orthofinder"]] <- all_data[["orthofinder"]] %>%
  left_join(cogeqc_outputs[["og_hscores_scaled_all"]], by = c("Orthogroup" = "Orthogroup"))

  all_data[["broccoli"]] <- all_data[["broccoli"]] %>%
  left_join(cogeqc_outputs[["og_hscores_scaled_all"]], by = c("Orthogroup" = "Orthogroup"))
  
  all_data_ogs <- all_data %>%
    bind_rows() %>%
    group_by(Orthogroup) %>%
    summarise(
      seq = paste(unique(seq), collapse = ","),
      parent_seq = paste(unique(parent_seq), collapse = ","),
      clean_seq = paste(unique(clean_seq), collapse = ","),
      clean_parent_seq = paste(unique(clean_parent_seq), collapse = ","),
      id = paste(unique(id), collapse = ","),
      gene_family_name = paste(unique(gene_family_name), collapse = ","),
      hmm_acc = paste(unique(hmm_acc), collapse = ","),
      hmm_name = paste(unique(hmm_name), collapse = ","),
      clan = paste(unique(clan), collapse = ","),
      supergroup = paste(unique(supergroup), collapse = ","),
      Score = paste(unique(Score), collapse = ","),
      Source = paste(unique(Source), collapse = ","),
      Score_scaled = paste(unique(Score_scaled), collapse = ",")
    )

saveRDS(all_data, all_data_path)
}



```

---

## Algorithm Comparison
```{r Algorithm Comparison}
# Create lists of sequences associated with each algorithm
compare_callers <- list()
compare_callers[["orthofinder"]] <- split(all_data[["orthofinder"]][["clean_seq"]], all_data[["orthofinder"]][["Orthogroup"]])
compare_callers[["orthofinder"]] <- lapply(compare_callers[["orthofinder"]], unique)

compare_callers[["broccoli"]] <- split(all_data[["broccoli"]][["clean_seq"]], all_data[["broccoli"]][["Orthogroup"]])
compare_callers[["broccoli"]] <- lapply(compare_callers[["broccoli"]], unique)

# Initialize an empty list to store the Jaccard similarity matrices
jaccard_results <- list()

# Perform a comparison and store the result in the list with a meaningful name
jaccard_results$of_vs_br <- calculate_jaccard_similarity(compare_callers[["orthofinder"]], compare_callers[["broccoli"]])
jaccard_results$of_vs_of <- calculate_jaccard_similarity(compare_callers[["orthofinder"]], compare_callers[["orthofinder"]])
jaccard_results$br_vs_br <- calculate_jaccard_similarity(compare_callers[["broccoli"]], compare_callers[["broccoli"]])

# Assuming jaccard_results is a list with matrices for each pair
metrics_of_vs_br <- calculate_jaccard_metrics(jaccard_results$of_vs_br, "orthofinder", "broccoli")
metrics_br_vs_br <- calculate_jaccard_metrics(jaccard_results$br_vs_br, "broccoli", "broccoli")
metrics_of_vs_of <- calculate_jaccard_metrics(jaccard_results$of_vs_of, "orthofinder", "orthofinder")

metrics_all <- rbind(metrics_of_vs_br, metrics_br_vs_br, metrics_of_vs_of)

metrics_all_long <- metrics_all %>%
  gather(key = "Metric", value = "Value", -tool1, -tool2)

# Create the heatmap using ggplot2
ggplot(metrics_all_long, aes(x = tool1, y = tool2, fill = Value)) +
  geom_tile() +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", midpoint = 0.5) + 
  facet_wrap(~Metric, scales = "free") +
  theme_minimal() +
  labs(
    title = "Heatmaps of Metrics by Tool Pair",
    x = "Tool 1", y = "Tool 2", fill = "Value"
  ) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.text = element_text(size = 12))


```

```{r Taxonomic Distribution}
# Example usage with your data frame (df_main)
upset_input <- list()
upset_input$orthofinder <- create_presence_matrix(all_data[["orthofinder"]], supergroup, Orthogroup)
upset_input$broccoli <- create_presence_matrix(all_data[["broccoli"]], supergroup, Orthogroup)

upset_input_all <- upset_input %>% bind_rows() # flatten the 

# join in metadata to maybe use in plot
 
upset_input_all <- upset_input_all %>% 
    select(Orthogroup, Source) %>%
    left_join(upset_input_all, by = c("Orthogroup" = "Name"))

supergroups <- c("Choanoflagellata", "Cnidaria", "Ctenophora", "Deuterostomia", "Filasterea", "Porifera", "Protostomia")


  

upset(upset_input_all,
  supergroups,
  annotations = list(
      'Orthogroup caller'=(
          ggplot(mapping=aes(fill=Source))
          + geom_bar(stat='count', position='fill')
          + scale_y_continuous(labels=scales::percent_format())
          + scale_fill_manual(values=c(
              'orthofinder'='#E41A1C', 'broccoli'='#377EB8'
          ))
          + ylab('Orthogroup caller')
      )
  ),
)


```